### PR TITLE
Replace Travis/AppVeyor with GitHub Actions

### PR DESCRIPTION
**Pull request by niik**
:wave: We, the GitHub Desktop Team, are happy users of prebuild in several projects that we depend upon. We've recently been working on yet another native module and ran into trouble building using the `windows-latest` GitHub Actions image. This seems to be resolved in a more recent version of node-gyp which is how I stumbled upon https://github.com/prebuild/prebuild/issues/286 and more specifically https://github.com/prebuild/prebuild/issues/286#issuecomment-1049305694

> Replace Travis (and Appveyor) with a GitHub Actions workflow, because our CI is currently broken and we'd like to test all of this. Also replace the badges in the README.

I figured we (I) could spend some time on this to see how difficult it would be. Overall the end result isn't too bad, macOS and Linux builds are both passing and Windows builds only have two failing tests, both related to nw-gyp.

The "can prebuild a gyp native module for node-webkit" test fails with the following error

```
Traceback (most recent call last):
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\gyp_main.py", line 16, in <module>
          sys.exit(gyp.script_main())
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/__init__.py", line 545, in script_main
          return main(sys.argv[1:])
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/__init__.py", line 538, in main
          return gyp_main(args)
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/__init__.py", line 514, in gyp_main
          options.duplicate_basename_check)
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/__init__.py", line 98, in Load
          generator.CalculateVariables(default_variables, params)
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/generator/msvs.py", line 1916, in CalculateVariables
          generator_flags.get('msvs_version', 'auto'))
        File "D:\a\prebuild\prebuild\node_modules\nw-gyp\gyp\pylib/gyp/MSVSVersion.py", line 444, in SelectVisualStudioVersion
          versions = _DetectVisualStudioVersions(version_map[version], 'e' in version)
      KeyError: '2019'
```

This is due to `nw-gyp` [using an old version of gyp](https://github.com/nwjs/nw-gyp/blob/1eb5e1b0b45f7257bebd46d80877c915a5b77bb5/gyp/pylib/gyp/MSVSVersion.py#L421-L435) which doesn't have an entry for VS 2019. Unfortunately that project seems rather stale so I don't know what the best way forward here is.

I also had to bump cmake-js in order to support VS 2019 (I chose the latest version but it's possible we would only need [6.1.0](https://github.com/cmake-js/cmake-js/releases/tag/v6.1.0)).

**TL;DR** A GitHub workflow for buillding on Windows (currenty failing), macOS (passing), and Linux (passing). Hoping to get some help figuring out the best path forward to get tests passing on Windows with VS2019.

### CODE REVIEW

**Comment by vweevers** [comment]
> This is due to nw-gyp [using an old version of gyp](https://github.com/nwjs/nw-gyp/blob/1eb5e1b0b45f7257bebd46d80877c915a5b77bb5/gyp/pylib/gyp/MSVSVersion.py#L421-L435) which doesn't have an entry for VS 2019. Unfortunately that project seems rather stale so I don't know what the best way forward here is.

Seems it was fixed: https://github.com/nwjs/nw-gyp/commit/966ca060299ea3a31ccd81b35916ccb26980faed

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: recursive
+        # Needed to fix node-webkit prebuilds, see https://github.com/nwjs/node/issues/46
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        if: ${{ startsWith(matrix.os, 'ubuntu') }}
+        with:
+          python-version: 2.7
```
[the remark this answers is not in the document]
Comment (verhovsky) [reply .github/workflows/ci.yml:42]: prebuild currently doesn't work with Python 3.11 because it uses an old version of node-gyp. Python 3.10 should work.

https://github.com/prebuild/prebuild/issues/286

**Comment by niik** [comment]
> @niik Thank you for this PR. Is there anything else you wanted to add or change before merging? I've left a couple of questions/comments inline, primarily around newer versions released since this was first opened that we may now want to take advantage of.

:wave: I applied your suggestions and the good news is that Windows builds are now passing, the bad news is that macOS and Linux builds are now failing. See https://github.com/niik/prebuild/actions/runs/4912267923/jobs/8771095211

I'm afraid I don't have much bandwidth to dig into this so hopefully there's something in here that y'all can figure out. For the macOS builds the error seems pretty clear

```
prebuild info begin Prebuild version 11.0.4
prebuild info build Preparing to prebuild native@1.0.0 for node-webkit 0.26.6 on linux-x64 using nw-gyp
prebuild ERR! configure error 
prebuild ERR! stack Error: Python executable "/opt/hostedtoolcache/Python/3.10.11/x64/bin/python" is v3.10.11, which is not supported by gyp.
```

The Node 20 builds are failing for different reasons it seems.

Let me know if there's anything I can do to help but also feel free to take this PR as-is and iterate on it in subsequent PRs.

**Comment by verhovsky** [comment]
Upgrading node-gyp should fix the macOS build error. The current version definitely doesn't work on Python 3.11 but seems to works fine if manually upgraded #286

**Comment by lovell** [comment]
It looks like that failing test is due to a (documented) lack of support for Python 3 in the `nw-gyp` dependency (a fork of an old version of `node-gyp`, has ~8k downloads/week, I suspect mostly driven by being a dependency of this module).

It's possible to install both Python 2 and 3 via `actions/setup-python`:

https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#specifying-multiple-pythonpypy-version

If this works, then the addition of a `--python=python2` flag here should help.

https://github.com/prebuild/prebuild/blob/e148f4554fa1fa435b79f5964c5fc06c9bda1331/test/native-module-gyp/package.json#L8

Alternatively we might consider skipping this test until `nw-gyp` supports Python 3.

**Comment by lovell** [comment]
Applying commit https://github.com/lovell/prebuild/commit/4f504140fb666e32063d7b83bb659f42b86c39f7 to the `niik:aaaaaand-action-(s)` branch appears to get the GitHub Actions based CI passing, e.g. https://github.com/lovell/prebuild/actions/runs/5169702326

* The `nw-gyp` dependency has a `--python` flag to specify the version but sadly it doesn't appear to work as advertised so we'll have to skip this test until https://github.com/nwjs/nw-gyp/issues/122 is fixed.

* I had to remove Node.js 20 for now as it (correctly) makes the `process.config.target_defaults` object read-only but node-gyp v6 tries to modify it, which produces the following error:
  > TypeError: Cannot assign to read only property 'cflags' of object '#<Object>'
  > at createConfigFile (D:\a\prebuild\prebuild\node_modules\node-gyp\lib\configure.js:118:21)

**Comment by verhovsky** [comment]
please merge this so that someone can update the node-gyp, I'm wasting so much time on this

**Comment by lovell** [comment]
I've pushed my change to the `niik:aaaaaand-action-(s)` branch via commit https://github.com/niik/prebuild/commit/bfeab27458f999f92fd4ec12312724a484917089 but it looks like there are temporary networking problems with GitHub Actions at the moment so I'll try again later - see https://github.com/niik/prebuild/actions/runs/5452106212

**Comment by lovell** [comment]
I can't re-run Actions on the `niik:aaaaaand-action-(s)` branch so I pushed the same changes to the `lovell:aaaaaand-action-(s)` branch this morning and it passed - see https://github.com/lovell/prebuild/actions/runs/5462011420 so I think we're ready to :shipit:

**Review by lovell — APPROVED** [review]

**Comment by niik** [comment]
> I can't re-run Actions on the `niik:aaaaaand-action-(s)`

Just triggered a re-run for ya :+1:

**Comment by vweevers** [comment]
@lovell Go ahead and merge when you're happy with it (I personally don't have time to review and other folks are quiet).

**Comment by lovell** [comment]
Thank you very much @niik and everyone else that has helped with this.

**Comment by lovell** [comment]
I've created a follow-up PR for this at https://github.com/prebuild/prebuild/pull/304
